### PR TITLE
Lsp server

### PR DESCRIPTION
**Pull request by rayakoren**

### CODE REVIEW

File: packages/xml-lsp-server/package.json [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    "@typescript-eslint/parser": "2.19.0",
+    "@typescript-eslint/eslint-plugin": "2.19.0",
+    "eslint": "6.8.0",
+    "eslint-config-prettier": "6.10.0",
+    "eslint-plugin-prettier": "3.1.2",
+    "source-map-support": "0.5.16",
+    "ts-node": "8.6.2"
+  },
+  "scripts": {
+    "ci": "npm-run-all lint coverage",
```
Comment (bd82) [inline packages/xml-lsp-server/package.json]: ci should also compile, otherwise the release process to npmjs.com
will have no artifacts included...

File: packages/xml-lsp-server/package.json [hunk 2]
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "@xml-tools/xml-lsp-server",
+  "description": "XML language server",
+  "version": "1.0.0",
+  "license": "Apache-2.0",
+  "engines": {
+    "node": ">=10.0.0"
+  },
+  "repository": {
```
Comment (bd82) [inline packages/xml-lsp-server/package.json]: a main property is needed to expose APIs to consumers

Comment (bd82) [inline packages/xml-lsp-server/package.json]: a `files` section property may be needed to expose generated .js artifacts in npmjs.com
you can check this by running `yarn pack` and examining the tar file

File: packages/xml-lsp-server/package.json [hunk 3]
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "@xml-tools/xml-lsp-server",
```
Comment (bd82) [inline packages/xml-lsp-server/package.json]: the `xml-` in the package name is redundant as we have `@xml-tools` scope

File: packages/xml-lsp-server/src/server.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import {
+  createConnection,
+  TextDocuments,
+  TextDocumentChangeEvent,
+  ProposedFeatures
+} from "vscode-languageserver";
+import { validateDocument } from "./serverHelper";
```
Comment (bd82) [inline packages/xml-lsp-server/src/server.ts]: I believe `serverHelper` could not be called `language-services`

File: packages/xml-lsp-server/src/serverHelper.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
+import {
```
Comment (bd82) [inline packages/xml-lsp-server/src/serverHelper.ts]: this repo does not use camel casing for file names.
use all lower characters with dashes instead.
`server-helper.ts`

File: packages/xml-lsp-server/src/serverHelper.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  message: error.message,
+  range: Range.create(
+    document.positionAt(error.offset),
+    document.positionAt(error.offset + error.length)
+  ),
+  severity: DiagnosticSeverity.Error,
+  source: SYNTAX_ERROR_MSG
+});
+
+const parsingErrorToDiagnostic = (document: TextDocument) => (
```
Comment (bd82) [inline packages/xml-lsp-server/src/serverHelper.ts]: `parsingErrorToDiagnostic` is a function, declare it directly as such
no need to assign a function Object to a const variable.

File: packages/xml-lsp-server/src/serverHelper.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  DiagnosticSeverity,
+  Range
+} from "vscode-languageserver";
+
+import { ILexingError, IRecognitionException } from "chevrotain";
+import { parse } from "@xml-tools/parser";
+
+export const SYNTAX_ERROR_MSG = "Syntax error";
+
+const lexingErrorToDiagnostic = (document: TextDocument) => (
```
Comment (bd82) [inline packages/xml-lsp-server/src/serverHelper.ts]: `lexingErrorToDiagnostic` is a function, declare it directly as such
no need to assign a function Object to a const variable.

File: packages/xml-lsp-server/src/serverHelper.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+});
+
+export async function validateDocument(
+  document: TextDocument
+): Promise<Diagnostic[]> {
+  let diagnostics: Diagnostic[] = [];
+  if (document.languageId === "xml") {
+    const { lexErrors, parseErrors } = parse(document.getText());
+    diagnostics = [
+      ...lexErrors.map(lexingErrorToDiagnostic(document)),
```
Comment (bd82) [inline packages/xml-lsp-server/src/serverHelper.ts]: nice usage of new `...` ESNext syntax 👍

File: packages/xml-lsp-server/test/server.spec.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
+import { expect } from "chai";
+
+import { validateDocument, SYNTAX_ERROR_MSG } from "../src/serverHelper";
```
Comment (bd82) [inline packages/xml-lsp-server/test/server.spec.ts]: really really minor, but generally you want to first import
external deps and only then relative deps `../src/serverHelper`

File: packages/xml-lsp-server/test/server.spec.ts [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+import {
+  TextDocument,
+  Diagnostic,
+  DiagnosticSeverity,
+  Position
+} from "vscode-languageserver";
+
+describe("Test XML Language Server", () => {
```
Comment (bd82) [inline packages/xml-lsp-server/test/server.spec.ts]: All the `describe` and `it` need text in BDD style.
- https://mochajs.org/#bdd
- [Background Info](https://en.wikipedia.org/wiki/Behavior-driven_development#Story_versus_specification)

```javascript
describe("the XML Language Server")
  it("will detect a parsing error in an xml document")
  it("will detect a lexing error in an xml document")
  it("will not detect any error in a none XML document")
```

File: packages/xml-lsp-server/test/server.spec.ts [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+  it("Test parsing error validation in xml document", async () => {
+    const doc: TextDocument = createTextDocument("xml", ">");
+    const pos: Position = { line: 0, character: 0 };
+    const diagnostic: Diagnostic = {
```
Comment (bd82) [inline packages/xml-lsp-server/test/server.spec.ts]: `diagnostic` variable is redundant as confusing as we also have `diagnostic**s**`

I'd recommend to in-line it in:
```typescript
    const expectedDiagnostics: Diagnostic[] = [diagnostic];
```

File: packages/xml-lsp-server/test/server.spec.ts [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+      message: `Expecting token of type --> Name <-- but found --> '>' <--`,
+      range: {
+        start: pos,
+        end: pos
+      },
+      severity: DiagnosticSeverity.Error,
+      source: SYNTAX_ERROR_MSG
+    };
+
+    const expectedDiagnostics: Diagnostic[] = [diagnostic1, diagnostic2];
```
Comment (bd82) [inline packages/xml-lsp-server/test/server.spec.ts]: We do not need a reference to diagnostic1/2 and it is also similar to `diagnostic**s**` name I'd recommend in-lining these two into `expectedDiagnostics`

**Comment by bd82** [comment]
A reference in the root level readme to this new package should be added
- https://github.com/SAP/xml-tools (see existing references)

File: packages/xml-lsp-server/package.json (hunk 2 above)
Comment (rayakoren) [reply packages/xml-lsp-server/package.json]: not sure what did you mean, I've added: "main": "src/server.ts"

Comment (rayakoren) [reply packages/xml-lsp-server/package.json]: Don't understand what should be under "files" and why

File: packages/xml-lsp-server/src/serverHelper.ts (hunk 6 above)
Comment (rayakoren) [reply packages/xml-lsp-server/src/serverHelper.ts]: It's a map method, that maps a nested object, what's wrong with it being declared in this way?

File: packages/xml-lsp-server/src/serverHelper.ts (hunk 7 above)
Comment (rayakoren) [reply packages/xml-lsp-server/src/serverHelper.ts]: It's a map method, that maps a nested object, what's wrong with it being declared in this way?

File: packages/lsp-server/package.json [hunk 13]
```diff
@@ -1,11 +1,12 @@
 {
-  "name": "@xml-tools/xml-lsp-server",
+  "name": "@xml-tools/lsp-server",
   "description": "XML language server",
   "version": "1.0.0",
   "license": "Apache-2.0",
   "engines": {
     "node": ">=10.0.0"
   },
+  "main": "src/server.ts",
```
Comment (bd82) [inline packages/lsp-server/package.json]: - main is always a js file
- also I am not sure we want a require("@xml-tools/lsp-server) to automatically launch the server.
- I guess if we had a programmatic API to start the server it would appear here, but we do not yet.
- Perhaps our programmatic API should expose the **PATH** to the server file which is needed by the LSP Client?

File: packages/lsp-server/package/.eslintrc.json [hunk 14]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (bd82) [inline packages/lsp-server/package/.eslintrc.json]: what is this "package" subfolder?

File: packages/lsp-server/package/package.json [hunk 15]
```diff
@@ -0,0 +1,36 @@
+{
```
Comment (bd82) [inline packages/lsp-server/package/package.json]: should not this file sit directly under packages/lsp-server? why the additional `package` sub folder?

File: packages/lsp-server/package/.eslintrc.json (hunk 14 above)
Comment (rayakoren) [reply packages/lsp-server/package/.eslintrc.json]: shouldnt be there, i'll remove

File: packages/lsp-server/package/test/server.spec.ts [hunk 16]
```diff
@@ -0,0 +1,68 @@
+import { expect } from "chai";
```
Comment (bd82) [inline packages/lsp-server/package/test/server.spec.ts]: this file should probably be named `langauge-services.spec.ts`

File: packages/lsp-server/package.json (hunk 13 above)
Comment (rayakoren) [reply packages/lsp-server/package.json]: the path should be from the package published in npm.js no?

File: packages/lsp-server/package/test/server.spec.ts [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+import {
+  TextDocument,
+  Diagnostic,
+  DiagnosticSeverity,
+  Position
+} from "vscode-languageserver";
+
+import { validateDocument, SYNTAX_ERROR_MSG } from "../src/language-services";
+
+describe("the XML Language Server", () => {
```
Comment (bd82) [inline packages/lsp-server/package/test/server.spec.ts]: because of previous comment (file name)
`the XML Language Server` -->  `the XML Language Services`

File: packages/lsp-server/package.json (hunk 13 above)
Comment (bd82) [reply packages/lsp-server/package.json]: - we could also expose the langauge services too (without the server)

File: packages/xml-lsp-server/src/serverHelper.ts (hunk 6 above)
Comment (bd82) [reply packages/xml-lsp-server/src/serverHelper.ts]: - **Consistency**, why are some functions using `function` **keyword** and others are assigned to a variable.
- **Different runtime behavior:**
  - regular functions gets **hoisted** to the top of the file and can be invoked before their definition or even in a cyclic manner, sort of pre-processing of the js file phase.
  - functions references inside variables are defined at actual runtime phase and cannot access not yet defined other function references in variables.
  - The **fat arrow** functions actual also bind the "this" context:
    - for **fat arrow** it would be an empty object (but can be modified)
    - for regular functions it would be a **global** object.

```javascript
const foo = () => {
  bar();
};

// won't work (comment in to check)
// foo();

const bar = () => {
  console.log("hi")
};

// will work
foo();


function foo2() {
  bar2();
}

// will work
foo2();

function bar2() {
  console.log("hello")
}
```

```javascript
function foo(){
  console.log(this) // inspect `this` in debugger
}
foo();

var foo2 = () =>{
  console.log(this) // inspect `this` in debugger
};

foo2();

```

So basically we add complexity that requires in-depth knowledge of JS runtime to understand and can cause some potential class of bugs without any benefit or gain...

Such style of functions declarations is used when dynamically creating functions.
e.g: https://lodash.com/docs/4.17.15#partial

File: packages/xml-lsp-server/src/serverHelper.ts (hunk 7 above)
Comment (bd82) [reply packages/xml-lsp-server/src/serverHelper.ts]: see above
